runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -238,8 +226,25 @@ public void write(IToken token, Object data) throws OverwriteException {
     public @Nonnull
     ILogData read(long address) {
         if (!runtime.getParameters().isCacheDisabled()) {
+            // The VersionLockedObject and the Transaction layer will generate
+            // undoRecord(s) during a transaction commit, or object sync. These
+            // undo records are stored in transient fields and are not persisted.
+            // A missing undo record can cause a NoRollbackException, thus forcing
+            // a complete object rebuild that generates a "scanning" behavior
+            // which affects the LRU window. In essence, affecting other cache users
+            // and making the VersionLockedObject very sensitive to caching behavior.
+            // A concrete example of this would be unsynchronized readers/writes:
+            // 1. Thread A starts replicating write1
+            // 2. Thread B discovers the write (via stream tail query) and
+            //    tries to read write1
+            // 3. Thread B's read results in a cache miss and the reader thread
+            //    starts loading the value into the cache
+            // 4. Thread A completes its write and caches it with undo records
+            // 5. Thread B finishes loading and caches the loaded value replacing
+            //    the cached value from step 4 (i.e. loss of undo records computed
+            //    by thread A)
             try {
-                return readCache.get(address);
+                readCache.get(address, () -> fetch(address));
```

Comment:
Putting `fetch` inside `get()` might not be good. If this api blocks writer, writer can starve because `fetch` can really take a long time due to hole fills. If this api doesn't block writer, then the NoRollBackException issue will still occur quite often as this reader may override the writer's value with no undo record.

Reply:
I have the same question, wouldn't we still be overwriting the writers value? shouldn't it also go though the cacheLoadAndGet?

Reply:
Based on code inspection the puts and get/load are protected by a segment lock, but there is another issue related to concurrency. I'll remove the fetch from the load.

Reply:
Actually, one of the tests was failing because of this.